### ISSUE TITLE
[Microsoft.TypeScript.MSBuild] Incremental build is broken for .NET (Core)

### ISSUE THREAD

**Issue by czdietrich** [microsoft/TypeScript#50599]
Building a project that uses `Microsoft.TypeScript.MSBuild` NuGet using e.g. `dotnet build` will compile typescript files each run even though there were no changes.

NuGet version: 4.8.2

## Details

The target `PreComputeCompileTypeScriptWithTSConfig` will run before the actual typescript compilation and use the `GenerateOutputLogs` task to check whether there is already a valid cache file in the obj folder in the form of `Tsc<NUMBERS>.out`. Unfortunately this `<NUMBERS>` block is generated by using the `string.GetHashCode()` method on the file path of the passed `tsconfig.json` file.
This hash code function is not stable in .NET (Core) and will be different for each new process, see also _Remarks_ in the [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.string.gethashcode?view=net-6.0#system-string-gethashcode). That's why the normal hash code of a `string` cannot be used to persist data.

## Proposed Fix

There was a similar issue for the incremental build in the WPF repo. The fix was just to replace the hash code generation with a stable version. See also [this PR](https://github.com/dotnet/wpf/pull/4053/files).

_I could not find the code of the Microsoft.TypeScript.MSBuild NuGet in this repo else I had also made a PR containing the fix._

**Comment by groogiam** [comment]
Is there any update on this? This is killing productivity. We have typescript complication on a project that has some long running AOT and this causes every build during the development to rebuild this expensive project. Thanks.

**Comment by czdietrich** [comment]
Would be nice, if this is getting fixed. The patch is trivial.

_Since there was no activity on this issue we fixed it on ourselves by creating a copy of this NuGet using a stable string hashing._

**Comment by czdietrich** [comment]
Any chance we can just patch this issue? This is an extreme productivity stopper and just costs time and money. 
The fix is simple, so please let us invest these hand full of minutes and fix this in the official NuGet.

**Comment by georg-eckert-zeiss** [comment]
Please release a fix for this.

**Comment by joj** [comment]
Removed mine and assigned myself to look into this

**Comment by pombredanne** [comment]
@joj gentle ping, also would you know where is the source code for this? I could not find it anywhere.

**Comment by joj** [comment]
Thanks for the ping, this got lost in the noise for some reason. I'm creating an internal item to track it to make sure that we look into it. Sorry for that!

**Comment by joj** [comment]
On the source code, the typescript nuget code was never open sourced (like the rest of typescript). I can make the change, though.

**Comment by pombredanne** [comment]
@joj re:

> On the source code, the typescript nuget code was never open sourced (like the rest of typescript). I can make the change, though.

This would be super gentle and mightily useful to have this open source! IMHO, this would have made this issue mostly a non-issue as it could have been fixed with a PR a while back :)

**Comment by joj** [comment]
Yeah, I see your point. Making something open source is a very intentional decision, though, and at some point we decided the nuget didn't meet the bar. I'll try to get that change soonish... I was researching and the solution to the hash code problem people are coming up with I'm not convinced about. I'll try something better (more standard).

**Comment by czdietrich** [comment]
Hi, is there any progress on this issue?

**Comment by safacero** [comment]
I fixed this today, should be on next release :D

**Comment by czdietrich** [comment]
Is there a prerelease NuGet which we can test?

**Comment by czdietrich** [comment]
There was a beta release for 5.3.0 of that NuGet recently. 
We tested it on our side and can confirm that it now works as expected.
Thank you, the issue is now solved from my point of view, so we could close ticket.

**Comment by jeremysalwen** [comment]
I have tested rebuilding with `Microsoft.TypeScript.MSBuild` 5.6.2 and it still unnecessarily runs the `PreComputeCompileTypeScriptWithTSConfig ` step.  5.3.2 also has the same behavior:

 ![Image](https://github.com/user-attachments/assets/afa4ab7e-125f-40cf-8b66-96325903c5ec)

**Comment by dazbradbury** [comment]
We see the same thing, would be good to know if the `PreComputeCompileTypeScriptWithTSConfig` is required even when no `.ts` changes have been made. To date I've assumed this process is checking whether the files have changed and the TS compiler needs to run or not.

**Comment by joj** [comment]
PreCompute is a pre-process we run to check if we actually need to run the full thing. It runs always, and should be super fast.

**Comment by jeremysalwen** [comment]
In my rebuild, it is the #1 most expensive task, and I only have a single typescript file in the entire solution.

**Comment by icnocop** [comment]
This issue is marked with milestone `TypeScript 5.7.0`, but I'm still experiencing this issue with version `5.7.1` of the NuGet package `Microsoft.TypeScript.MSBuild`.

`PreComputeCompileTypeScriptWithTSConfig` and `CompileTypeScriptWithTSConfig` take about 62 - 65 seconds each.


`MyWebApplication.csproj`:
```xml
<TypeScriptToolsVersion>4.3</TypeScriptToolsVersion>
```

`tsconfig.json`:
```json
{
  "compileOnSave": true,
  "compilerOptions": {
    "baseUrl": "./",
    "noImplicitAny": false,
    "noEmitOnError": true,
    "removeComments": true,
    "sourceMap": true,
    "target": "es5",
    "module": "system",
    "moduleResolution": "node",
    "allowJs": false,
    "typeRoots": [
      "node_modules/@types/",
      "scripts/typings/"
    ],
    "lib": [
      "es6",
      "dom"
    ],
    "incremental": true
  },
  "exclude": [
    "node_modules",
    "wwwroot"
  ]
}
```

When targeting multiple .net frameworks and `BuildInParallel` is set to `false`, I expected the tasks to only run once during the first build, for the first platform, and then skipped for subsequent .net frameworks.
Subsequent builds should take incremental rules into consideration, and the `CompileTypeScriptWithTSConfig` task should be skipped if there were no changes to any of the TypeScript files.

Thank you.
